knowledge_engine/knowledge_engine/stat/collectors/user_behavior.py

```
@@ -0,0 +1,719 @@
+    kb_clause, kb_params = build_kb_in_clause(mfilter.kb_ids)
+    kb_where = f"AND resource_id {kb_clause}" if kb_clause else ""
+
+    rows = source_session.execute(
+        text(f"""
+            SELECT resource_id AS kb_id, COUNT(DISTINCT user_id) AS analyst_count
+            FROM resource_members
+            WHERE role = 'RestrictedAnalyst'
+              AND resource_type = 'KnowledgeBase'
+              {kb_where}
+            GROUP BY resource_id
+        """),
```

Ruff (0.15.21)

[error] Possible SQL injection vector through string-based query construction

(S608)